0010-keccak-syscall.md

```diff
@@ -0,0 +1,402 @@
+
+<!-- DONE -->
+
+The register `v0` must be set to `1` and the register `a3` to `0`.
+
+### `sysExitGroup`
+
+<!-- DONE -->
+
+The program must halt in this case and `set_halted` must be called with the value `a0`.
+
+### `sysRead`
+
+This syscall is used to get data from the outside world and write in the VM memory.
+
+The register `a0` contains the file descriptor of the input channel, the
+register `a1` the address and the `a2` contains the number of bytes that must be
+read.
+
+At the end of the call, the register `v0` must contain the number of bytes that
+have been read, and `a3` must contain the error code.
+
+Different input channels must be supported.
+
+#### Standard input
+
+<!-- DONE -->
+
+In the case of the standard input, i.e. when `a0` equals `0`, nothing in
+particular happens. Therefore, the registers `v0` and `a3` are set to `0`, which
+corresponds to no data has been read and no error is raised.
+
+#### Pre-image oracle
+
+When the buffer to be read is the pre-image oracle, i.e. when `a0` equals `5`, a
+chunk of the pre-image is read, and copy into the memory.
+The pre-image that must be read is determined by the `preimageKey`, and the actual
+chunk is given by the preimage offset variable.
+
+The number of bytes that must be read is in the register `a2` and the address
+where the chunk as to be written is in the register `a1`.
+
+At the end of the call, the register `v0` must contain the number of bytes that
+have been read. Also, the preimage offset has to be incremented by the number of
+bytes that has been read.
+
+The preimage data length is located in the 8 first addresses, encoded on 64
+bits in big-endian.
+
+A difference with Cannon is we verify the preimage data is the actual data the
+user has requested previously. To achieve this, we will use the optimised
+keccak gate described in [Keccak RFC](./keccak.md).
+
+The keccak sponge state is 136 bytes, and is located in the columns `[100,
+168[`. While reading the preimage data, we can fill at the same time the sponge
+state. When 136 bytes have been read, a keccak permutation is executed.
+If we have finished reading the data, we run the permutation. In the case the total
+length is a multiple of 136, an additional permutation is executed
+with the expected padding.
+
+Reading additional bytes after the EOF must not have any effect and must be
+allowed. This can be achieved by activating the "preimage read gate" iff the
+different between the number of bytes that have to be read and the preimage
+offset is not zero. In the case we arrived at the end of the file, the preimage
+offset must not be increased and the number of bytes read must be set `0`.
+
+After we read the data and run the permutation a last time as described above,
+we must verify that the preimage key corresponds to the output of sponge.
+To achieve this, we compare the 31 bytes first bytes as the first byte is
+dropped for the preimage key type. The output of the keccak sponge is described
```

Comment:
yeah, so basically where and how do you do that? You're saying you will store the 31 bits of the key in that table and check that the (chopped) output of the keccak appears in that table? But then how do you link this one to a particular entry of the table? I mean, a normal lookup contains many terms in the table, and it solely gives you the guarantee that some term IS in the table. But doesn't tell you which entry. Here you not only want to check that you saw that (chopped) hash before as a file name, but you want to link it with the one you're just reading. How is this chain of connexions supposed to take place?